Steps to use these functions

rachelmyers opened the pull request:
I omitted usage instructions from the README; that was my mistake. This PR adds instructions to include the wipeout and takeout functions in a new project.

Because the functions cover wipeout and takeout for the Realtime Database, Firestore, and Storage, this also includes instructions for deleting the functions you don't need, though there's no harm in leaving them in.

💡

eobrain commented on README.md:
```diff
@@ -7,11 +7,53 @@ and copying out data at when a user requests it (the [`takeout` function]()).
+The developer specifies the paths to data to wipeout or takeout. Those paths live in `user_privacy.json`. The data structures vary for each of the products:
+- For the RTDB, it’s a list of Strings to the path in the database of the form `"</users/uid/follows/..."`
+- For Firestore, it’s a list of Objects of the form `{ "collection": "admins",
+"doc": "UID_VARIABLE", "field": "email" }`.
```
For this snippet and the one on the next bullet, I suggest using a ``` codeblock because currently the snippets wrap, making them a bit hard to read.

eobrain commented on README.md:
```diff
@@ -7,11 +7,53 @@ and copying out data at when a user requests it (the [`takeout` function]()).
+how to wire it up.
+
+The developer specifies the paths to data to wipeout or takeout. Those paths live in `user_privacy.json`. The data structures vary for each of the products:
+- For the RTDB, it’s a list of Strings to the path in the database of the form `"</users/uid/follows/..."`
```
Suggest following [Google markdown style guide][1] which favors `*` over `-`, and calls for 4-space indent.  Also I suggest wrapping at 80 chars everywhere.

[1]: https://github.com/google/styleguide/blob/gh-pages/docguide/style.md#lists

eobrain commented on README.md:
```diff
@@ -7,11 +7,53 @@ and copying out data at when a user requests it (the [`takeout` function]()).
 users and apps.
 
-### How to play with these functions
+### How to contribute
```
Is this really "contribute", as these instructions don't call for contributing back to this repo?  How about something like  "How to use these functions with your app"

ianbarber replied:
+1

rachelmyers replied:
Lines 10 - 17 are instructions for contributing back. The bulk of this PR is  instructions for using the functions in a new app starting below, at line 18.

ianbarber commented on README.md:
```diff
@@ -7,11 +7,53 @@ and copying out data at when a user requests it (the [`takeout` function]()).
+
+#### Wipeout function
+The wipeout function as written is triggered when a user deletes their account
+using firebase auth, and it performs a wipeout from all three services.
```
initial caps the products: Firebase Auth

ianbarber commented on README.md:
```diff
@@ -7,11 +7,53 @@ and copying out data at when a user requests it (the [`takeout` function]()).
+- [ ] Make sure you’ve added the ability for the user to delete their account, because that’s what triggers the function
+- [ ] If you’re not using all three products, RTDB, Firestore, and Storage, remove the parts of the function that you don’t need.
+
+As an FYI, the wipeout is implemented by collecting a promise for every deletion event that needs to occur. Only when all promises resolve is the wipeout considered complete. Keep that in mind if you send a confirmation message that wipeout has completed.
```
Drop "As an FYI, the " - its a valid note on its own.

ianbarber commented on README.md:
```diff
@@ -7,11 +7,53 @@ and copying out data at when a user requests it (the [`takeout` function]()).
+As an FYI, the wipeout is implemented by collecting a promise for every deletion event that needs to occur. Only when all promises resolve is the wipeout considered complete. Keep that in mind if you send a confirmation message that wipeout has completed.
+
+#### Takeout function
+The takeout function is triggered via a HTTP request. (The sample app in
```
Remove parens, I think that's a good note.

ianbarber commented on README.md:
```diff
@@ -7,11 +7,53 @@ and copying out data at when a user requests it (the [`takeout` function]()).
+  - [ ] A ``"takeoutUploadBucket"`` key that maps to the name of your primary bucket (or in the case of the free tier, your only bucket).
+  - [ ] Paths to personal information for all the products you’re using.
+- [ ] Wire it up in a way appropriate for your platform. For web, I added a
+button in `public/index.html`, that triggers a POST request, and add a rewrite entry in `firebase.json` to trigger the function via HTTP.
```
I would maybe split out the "For web" bit, simply the instruction, and then put a couple of lines on "how this looks in the sample" as a demonstration.

ianbarber commented on README.md:
```diff
@@ -7,11 +7,53 @@ and copying out data at when a user requests it (the [`takeout` function]()).
+- [ ] If you’re not using all three products, RTDB, Firestore, and Storage,
+remove the parts of the function you don’t need.
+
+Takeout for Storage is weird because we copy the data to a folder for that user, and also create the json of references.
```
Maybe rewrite this a little bit to touch on the expectations. 
"For Realtime Database and Firestore we write the user data into a JSON document. For Storage we write a JSON document containing an index of stored files, and copy the files themselves into a folder"

ianbarber commented on README.md:
```diff
@@ -7,11 +7,53 @@ and copying out data at when a user requests it (the [`takeout` function]()).
+
+Takeout for Storage is weird because we copy the data to a folder for that user, and also create the json of references.
+
+The storage.rules additions are extremely important for takeout. Without them, all the takeout data is freely available. We next the final takeout output under `/takeout`, but check to make sure you don’t have have wildcards in your storage rules, that will also apply to the takeout section.
```
Not sure what you're saying with "We next". If this can be a clearer steps (1. add these rules 2. check for wildcards) that might help?

rachelmyers replied:
Good catch; that looks like editing gone wrong.

rachelmyers replied:
Replacing with:
> The storage.rules additions are extremely important for takeout; without them, the takeout data is broadly available. By default, the takeout function uploads data to a top level /takeout folder, and the Storage Rules restrict access to the specific user. Be sure to restrict any preexisting rules with wildcards that would grant broader access to takeout data.

rachelmyers commented:
I addressed the specific changes; let me know if you see anything else that should change. I also moved the Contributing steps below the Usage steps, since most people probably want the Usage steps.
